presto-druid/src/main/java/com/facebook/presto/druid/segment/DruidSegmentReader.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.druid.segment;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.druid.DruidColumnHandle;
+import com.facebook.presto.druid.column.ColumnReader;
+import com.facebook.presto.druid.column.SimpleReadableOffset;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.PrestoException;
+import com.google.common.collect.ImmutableMap;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.QueryableIndex;
+import org.apache.druid.segment.column.BaseColumn;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.druid.DruidErrorCode.DRUID_SEGMENT_LOAD_ERROR;
+import static com.facebook.presto.druid.column.ColumnReader.createColumnReader;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+
+public class DruidSegmentReader
+        implements SegmentReader
+{
+    private static final int BATCH_SIZE = 1024;
+
+    private final Map<String, ColumnReader> columnValueSelectors;
+    private final long totalRowCount;
+
+    private QueryableIndex queryableIndex;
+    private long currentPosition;
+    private int currentBatchSize;
+
+    public DruidSegmentReader(SegmentIndexSource segmentIndexSource, List<ColumnHandle> columns)
+    {
+        try {
+            queryableIndex = segmentIndexSource.loadIndex(columns);
+            totalRowCount = queryableIndex.getNumRows();
+            ImmutableMap.Builder<String, ColumnReader> selectorsBuilder = ImmutableMap.builder();
+            for (ColumnHandle column : columns) {
+                DruidColumnHandle druidColumn = (DruidColumnHandle) column;
+                String columnName = druidColumn.getColumnName();
+                Type type = druidColumn.getColumnType();
+                BaseColumn baseColumn = queryableIndex.getColumnHolder(columnName).getColumn();
+                ColumnValueSelector<?> valueSelector = baseColumn.makeColumnValueSelector(new SimpleReadableOffset());
+                selectorsBuilder.put(columnName, createColumnReader(type, valueSelector));
+            }
+            columnValueSelectors = selectorsBuilder.build();
+        }
+        catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)